### PR TITLE
[arci-ros2] Improve subscription helper

### PR DESCRIPTION
**Pull request by taiki-e**

### CODE REVIEW

File: arci-ros2/src/utils.rs [hunk 1]
```diff
@@ -37,18 +24,45 @@ pub(crate) async fn wait(is_done: Arc<AtomicBool>) {
...
+                .build()
+                .unwrap()
+                .block_on(async move {
+                    for _ in 0..10 {
+                        if let Some(v) = subscriber.next().now_or_never() {
+                            return v;
+                        }
+                        tokio::time::sleep(Duration::from_millis(100)).await;
+                    }
+                    None // timeout
```
Comment (taiki-e) [inline arci-ros2/src/utils.rs:60]: I'm not sure how many seconds is a reasonable timeout, but I set it to 1s (100ms*10) for now.

File: arci-ros2/src/ros2_laser_scan.rs [hunk 2]
```diff
@@ -6,36 +8,35 @@ use crate::{utils, Node};
...
     /// Creates a new `Ros2LaserScan2D` from sensor_msgs/LaserScan topic name.
-    pub fn new(node: Node, laser_scan_topic_name: &str) -> Self {
-        Self {
-            node,
+    pub fn new(node: Node, laser_scan_topic_name: &str) -> Result<Self, Error> {
+        let mut scan_subscriber = node
+            .r2r()
+            .subscribe::<LaserScan>(laser_scan_topic_name, QosProfile::default())
+            .map_err(anyhow::Error::from)?;
+        let scan = utils::subscribe_one(&mut scan_subscriber);
```
[the remark this answers is not in the document]
Comment (taiki-e) [reply arci-ros2/src/ros2_laser_scan.rs:24]: > Ros2LaserScan2D::new() will lock(wait) until one message has come, right?

The current implementation waits until it times out in `new` (`subscribe_one `) and then attempts to subscribe in the spawned thread (`subscribe_thread`) even if the value has not yet been obtained.
The "failed to get ..." error only comes if the value has not yet been obtained (in the spawned thread) when current_scan is actually called.

AFAIK the behavior of the error here is the same as in previous implementations and in arci-ros's Localization, but can be changed if it is better to have the error occur in `new`.

[the remark this answers is not in the document]
Comment (OTL) [reply arci-ros2/src/ros2_laser_scan.rs]: I see. Thank you for your kind explanation!

File: arci-ros2/src/utils.rs [hunk 3]
```diff
@@ -37,18 +24,49 @@ pub(crate) async fn wait(is_done: Arc<AtomicBool>) {
...
+                .block_on(async move {
+                    let start = Instant::now();
+                    loop {
+                        if let Some(v) = subscriber.next().now_or_never() {
+                            return v;
+                        }
+                        if start.elapsed() > timeout {
+                            return None; // timeout
+                        }
+                        tokio::time::sleep(Duration::from_millis(100)).await;
```
Comment (OTL) [inline arci-ros2/src/utils.rs:63]: It might be better to add a TODO comment for 100ms as the minimum sleep duration, but it's a super small issue.

**Review by OTL — APPROVED** [review]

**Pull request merged**